### ISSUE TITLE
Including occlusion in annotations?

### ISSUE THREAD

**Issue by henrytkirk** [apple/turicreate#988]
Hello!
I'm working on an object detection model and I have a question about including items labeled as occlusion. How do I incorporate them into my model? It doesn't seem like the annotations are picking up the field I added.

Is there a way to train the model twice? (the docs aren't clear if you can take am model and train more data against it)

Thanks!

**Comment by henrytkirk** [comment]
Any thoughts to my question?

**Comment by gustavla** [comment]
Hi @henrytkirk. Really sorry for the long delay!

Any unrecognized field added to the annotation dictionary will be ignored by the model. Let me present some alternatives, but I will first make some guesses as to what you are trying to achieve. My guess is that you have for instance N classes and for certain objects you want to mark if the instance is occluded or not. Now, we do not have support for multi-label detection, which would mean having separately independent targets (1. class, 2. occlusion yes/no).

However, you can pre-combine these two targets, so that your new "classes" are:
* class1
* class1-occluded
* class2
* class2-occluded
* ...

This is not quite ideal, but it may work out.

As for your second question of training the model twice. I assume you mean starting the training from an already trained model, right? No, this is currently not supported. It's related to checkpointing (#574), although that feature request does not explicitly specify if it should be possible to resume training. This is something that we have discussed, but we have not committed to supporting it at this point. It introduces a lot of reproducibility issues and it is also not trivial how to set learning rates fully automatically in such a setting.

**Comment by henrytkirk** [comment]
Thanks for the detailed response! This is very helpful. The reason I asked about occlusion is because of a statement from the object detection user guide:

"If you are trying to detect persons, then every occurrence of a person should warrant a ground truth bounding box. If you leave some persons unmarked, the model can get confused why some persons are marked as positives (presence of an instance) while others are marked as negatives (absence of an instance)" (https://github.com/apple/turicreate/tree/master/userguide/object_detection)

So, I assumed that occluded items need to be considered. For example, what if my class was an Apple. And I had a bowl of Apples in an image. Do I simply put bounding boxes around the Apples that are fully visible? Or all of them including those partially visible? Or do I ignore these types of images for training purposes.

And to clarify my "train twice" question,  what I'm asking is if there's a way to use my existing model for transfer learning expand it's capabilities. It seems now that I have to start from scratch each time. Is that correct?

Thanks very much for your assistance. I really appreciate your time.

**Comment by gustavla** [comment]
Ah, I'm glad you elaborated on what you are trying to do. So, what that refers to is that you should not be inconsistent in what you annotate as samples. However, if you say for instance that you don't want partially occluded objects to be detected at all, that is a consistent behavior that the model can learn. With that said, you have many choices of what to do with occluded objects.

1. Ignore the occlusion and define the bounding box as if the object was fully visible.

2. Take the occlusion into account when defining the tight bounding box (this is popular in the academic literature).

3. Choose not detect partially visible objects at all.

All these three are consistent behaviors that the model can learn. The choice is entirely up to what you want the detector's behavior to be.

> And to clarify my "train twice" question, what I'm asking is if there's a way to use my existing model for transfer learning expand it's capabilities. It seems now that I have to start from scratch each time. Is that correct?

Yes, this is correct. We do not have support for this currently.

**Comment by henrytkirk** [comment]
Thank you VERY much for the detailed replies!
